### PR TITLE
[buteo-sync-plugins-social] Guard notebook pointer access. Contributes to MER#1173

### PR DESCRIPTION
**Pull request by chriadam**
This commit ensures that we check the value of the notebook pointer
returned from the lookup function after asynchronous network requests
have finished, otherwise (if the local notebook is deleted during sync)
dereferencing this pointer can cause a crash.

Contributes to MER#1173

### CODE REVIEW

**Comment by chriadam** [comment]
@pvuorela

File: src/google/google-calendars/googlecalendarsyncadaptor.cpp [hunk 1]
```diff
@@ -1342,6 +1342,11 @@ void GoogleCalendarSyncAdaptor::finishedRequestingRemoteEvents(int accountId, co
     foreach (const QString &updatedCalendarId, m_calendarsFinishedRequested.keys()) {
```
Comment (pvuorela) [inline src/google/google-calendars/googlecalendarsyncadaptor.cpp]: m_calendarsFinishedRequested seems to be used only inside this method and never cleared. Any chance bad data could accumulate?

**Comment by pvuorela** [comment]
LGTM

File: src/google/google-calendars/googlecalendarsyncadaptor.cpp (hunk 1 above)
Comment (chriadam) [inline src/google/google-calendars/googlecalendarsyncadaptor.cpp]: I don't think so - the plugin instance lifetime is exactly one sync with exactly one account, so unless there is a logic bug in the plugin which makes us request a single calendar multiple times, for example, I don't think the map could contain bad data. 

What it may contain, however, is out-of-date data, if the notebook is deleted, or its pluginName field is modified, between when we requested the calendar data, and this point in the code - this guard should now protect from that case.

**Comment by chriadam** [comment]
With a bit more thought... I wonder how the crash could be hit, in reality?

The notebook should only ever be deleted by the sync plugin itself (reacting to the account deletion).  But if that occurs during a sync, I believe that it is queued until the original sync has completed.

Perhaps I'm wrong about that... I will have to test that.  I hold off on merging this PR until I come up with a conclusive answer.

**Comment by chriadam** [comment]
I added a QThread::sleep(10); before the call to notebookForCalendarId(), so that I could delete the account via the Settings UI just before that line is hit.  It didn't crash.  What happens is that Buteo actually aborts the ongoing sync:

SocialdButeoPlugin::abortSync:193 - info :  aborting sync with status: 5

Once the abort is complete, Buteo starts a new instance of the plugin and invokes the Purge() / cleanup() code.

In short, everything seems to work "properly".

I guess that this PR is still "good" in that it protects from crashes in the case where the notebook is somehow deleted by some other process - but I don't think it's related to the crash which was reported in JB#30870.

**Comment by chriadam** [comment]
I will merge this one, and we can see if anyone is able to reproduce the original issue with this patch applied or not.  Either way, I think this is a good defensive programming measure anyway.
